gmail/src/models/partner.ts

```diff
@@ -0,0 +1,215 @@
+            : null;
+        return partner;
+    }
+
+    static fromOdooResponse(values: any): Partner {
+        const partner = new Partner();
+
+        if (values.id && values.id > 0) {
+            partner.id = values.id;
+        }
+        partner.name = values.name;
+        partner.email = values.email;
+
+        partner.phone = values.phone;
+        partner.mobile = values.mobile;
+        partner.image = values.image ? "data:image/png;base64," + values.image : null;
+
+        if (values.company && values.company.id && values.company.id > 0) {
+            partner.company = Company.fromOdooResponse(values.company);
+        }
+
+        return partner;
+    }
+    /**
+     * Try to find information about the given email /name.
+     *
+     * If we are not logged to an Odoo database, enrich the email domain with IAP.
+     * Otherwise fetch the partner on the user database.
+     */
+    static enrich(email: string, name: string): [Partner, ErrorMessage] {
+        const odooServerUrl = State.odooServerUrl;
+        const odooAccessToken = State.accessToken;
+
+        return odooServerUrl && odooAccessToken ? this.getPartner(email, name) : this._enrichFromIap(email, name);
+    }
+
+    /**
+     * Extract the email domain and send a request to IAP
+     * to find information about the company.
+     */
+    static _enrichFromIap(email: string, name: string): [Partner, ErrorMessage] {
+        const odooSharedSecret = State.odooSharedSecret;
+        const userEmail = Session.getEffectiveUser().getEmail();
+
+        const senderDomain = email.split("@").pop();
+
+        const response = postJsonRpcCached(URLS.IAP_COMPANY_ENRICHMENT, {
+            email: userEmail,
+            domain: senderDomain,
+            secret: odooSharedSecret,
+        });
+
+        const error = new ErrorMessage();
+        if (!response) {
+            error.setError("http_error_iap");
+        } else if (response.error && response.error.length) {
+            error.setError(response.error);
+        }
+
+        const partner = new Partner();
+        partner.name = name;
+        partner.email = email;
+
+        if (response && response.name) {
+            partner.company = Company.fromIapResponse(response);
+        }
+
+        return [partner, error];
+    }
+
+    /**
+     * Create a "res.partner" with the given values in the Odoo database.
+     */
+    static savePartner(partnerValues: any): number {
+        const url = State.odooServerUrl + URLS.PARTNER_CREATE;
+        const accessToken = State.accessToken;
+
+        const response = postJsonRpc(url, partnerValues, {
+            Authorization: "Bearer " + accessToken,
+        });
+
+        return response && response.id;
+    }
+
+    /**
+     * Fetch the given partner on the Odoo database and return all information about him.
+     */
+    static getPartner(email: string, name: string, partnerId: number = null): [Partner, ErrorMessage] {
```

Comment:
getPartner will return only 1 partner, it can be used to retrieve the partner from his ID (to refresh the card, etc), but searchPartner will return a list of partner whose name / email match the query (to match the 2 endpoints we have on the Odoo side)